File: Data Structures/Assignment2/build_heap.py
# python3

class HeapBuilder:
    def __init__(self):
        self._swaps = []
        self._data = []

    def ReadData(self):
      n = int(input())
      self._data = [int(s) for s in input().split()]
      assert n == len(self._data)

    def WriteResponse(self):
        print(len(self._swaps))
        for swap in self._swaps:
            print(swap[0], swap[1])

    def parent(self, i):
        return int((i - 1) / 2)

    def left_child(self, i):
        return int((i * 2 + 1))

    def right_child(self, i):
        return int((i * 2 + 2))

    def shift_down(self, i):
        left_chi = self.left_child(i)
        right_chi = self.right_child(i)
        j = i
        if (left_chi < len(self._data) and self._data[left_chi] < self._data[j]):
            j = left_chi
        if (right_chi < len(self._data) and self._data[right_chi] < self._data[j]):
            j = right_chi
        if (i != j):
            self._swaps.append((i, j))
            self._data[i], self._data[j] = self._data[j], self._data[i]
            self.shift_down(j)

    def GenerateSwaps(self):
        N = len(self._data)
        for i in range(N//2, -1, -1):
            self.shift_down(i)

    def Solve(self):
        self.ReadData()
        self.GenerateSwaps()
        self.WriteResponse()

if __name__ == '__main__':
    heap_builder = HeapBuilder()
    heap_builder.Solve()

File: Data Structures/Assignment2/build_heap.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;
using std::cin;
using std::cout;
using std::swap;
using std::pair;
using std::make_pair;
using std::max;

class HeapBuilder {
private:
    vector<int> data_;
    vector< pair<int, int> > swaps_;
    
    void WriteResponse() const {
        cout << swaps_.size() << "\n";
        for (int i = 0; i < swaps_.size(); ++i) {
            cout << swaps_[i].first << " " << swaps_[i].second << "\n";
        }
    }
    
    void ReadData() {
        int n;
        cin >> n;
        data_.resize(n);
        for(int i = 0; i < n; ++i)
            cin >> data_[i];
    }
    
    int parent(int i) {
        if (i >0) {
            return (i - 1) / 2;
        }
        else {
            return i;
        }
    }
    
    int left_child(int i) { return 2 * i + 1; }
    
    int right_child(int i) { return 2 * i + 2; }
    
    void shit_up(int i) {
        while (i > 0 && data_[parent(i)] > data_[i]) {
            swap(data_[i], data_[parent(i)]);
            swaps_.push_back(make_pair(i, parent(i)));
            i = parent(i);
        }
    }
    
    void shift_down(int i) {
        int l = left_child(i);
        int r = right_child(i);
        int minIndex = i;
        if (l < data_.size() && data_[l] < data_[minIndex]) {
            minIndex = l;
        }
        if (r < data_.size() && data_[r] < data_[minIndex]) {
            minIndex = r;
        }
        if (i != minIndex) {
            swap(data_[i], data_[minIndex]);
            swaps_.push_back(make_pair(i, minIndex));
            shift_down(minIndex);
        }
    }
    
    void GenerateSwaps() {
        swaps_.clear();
        int size = (int)data_.size();
        for (int i = size/2; i >= 0; i--)  {
            shift_down(i);
        }
        
    }
    
public:
    void Solve() {
        ReadData();
        GenerateSwaps();
        WriteResponse();
    }
};

int main() {
    std::ios_base::sync_with_stdio(false);
    HeapBuilder heap_builder;
    heap_builder.Solve();
    return 0;
}
